core/src/connection_reuse.rs

```diff
@@ -40,205 +40,556 @@
+    fn reset_conn(self) {
+        if let PeerState::Active { closed, .. } = self {
+            closed.store(false, Ordering::Release);
+        }
+    }
 }
 
-impl<T, C> From<UpgradedNode<T, C>> for ConnectionReuse<T, C>
+/// Struct shared between most of the `ConnectionReuse` infrastructure.
+/// Knows about all connections and their current state, allows one to poll for
+/// incoming and outbound connections, while managing all state-transitions that
+/// might occur automatically.
+struct ConnectionsManager<T, C>
+where
+    T: Transport,
+    C: ConnectionUpgrade<T::Output, T::MultiaddrFuture>,
+    C::Output: StreamMuxer,
+{
+    /// Underlying transport and connection upgrade, used when we need to dial or listen.
+    transport: UpgradedNode<T, C>,
+
+    /// All the connections that were opened, whether successful and/or active or not.
+    // TODO: this will grow forever
+    connections: Mutex<FnvHashMap<Multiaddr, PeerState<C::Output>>>,
+
+    /// Tasks to notify when one or more new elements were added to `connections`.
+    notify_on_new_connec: Mutex<HashMap<usize, task::Task, UsizeHasherBuilder>>,
+
+    /// Counter giving us the next listener_id
+    listener_counter: Counter,
+}
+
+impl<T, C> ConnectionsManager<T, C>
 where
     T: Transport,
-    T::Output: AsyncRead + AsyncWrite,
     C: ConnectionUpgrade<T::Output, T::MultiaddrFuture>,
     C::Output: StreamMuxer,
+{
+    /// Notify all `Tasks` in `self.notify_on_new_connec` that a new connection was established
+    /// consume the tasks in that process
+    fn new_con_notify(&self) {
+        let mut notifiers = self.notify_on_new_connec.lock();
+        trace!("Notify {} listeners about a new connection", notifiers.len());
+        for (_, task) in notifiers.drain() {
+            task.notify();
+        }
+    }
+
+    /// Register `task` under id if not yet registered
+    fn register_notifier(&self, id: usize, task: task::Task) {
+        let mut notifiers = self.notify_on_new_connec.lock();
+        notifiers.insert(id, task);
+    }
+
+    /// Clear the cached connection if the entry contains an error. Returns whether an error was
+    /// found and has been removed.
+    fn clear_error(&self, addr: &Multiaddr) -> bool {
+        let mut conns = self.connections.lock();
+
+        if let Some(PeerState::Errored(ref err)) = conns.get(addr) {
+            trace!(
+                "Clearing existing connection to {} which errored earlier: {:?}",
+                addr,
+                err
+            );
+        } else {
+            return false;
+        }
+
+        // only reaching the point if the entry was an error
+        conns.remove(addr);
+        return true;
+    }
+}
+
+/// Polls the given outbound of addr for a new substream. Returns `Ok(Some(stream))` if available,
+/// `Ok(None)` if the stream is not ready yet add an `Err` if the stream had failed.
+fn poll_for_substream<O, S>(mut outbound: O, addr: &Multiaddr)
```

Comment:
Why isn't that a `Poll`, so that we return `Async::Ready` and `NotReady` instead of respectively `Some` and `None`?